includes/class-newspack-popups-segmentation.php

```diff
@@ -386,6 +408,68 @@ function( $segment ) {
 		return $segments;
 	}
 
+	/**
+	 * Get array of default segments by type.
+	 *
+	 * @return array Array of default segments.
+	 */
+	public static function get_default_segments() {
+		return get_option( self::DEFAULT_SEGMENTS_OPTION_NAME, false );
+	}
+
+	/**
+	 * Generate default segments. If they've already been generated, rebuild them with default parameters.
+	 *
+	 * @return array Array of all segments.
+	 */
+	public static function create_default_segments() {
+		$default_segments = get_option( self::DEFAULT_SEGMENTS_OPTION_NAME, [] );
+		$default_options  = [
+			'donors'           => [
+				'name'          => __( '[Newspack] Donors', 'newspack-popups' ),
+				'configuration' => [
+					'is_donor' => true,
+				],
+			],
+			'frequent_readers' => [
+				'name'          => __( '[Newspack] Frequent Readers', 'newspack-popups' ),
+				'configuration' => [
+					'min_posts' => 5,
+				],
+			],
+			'subscribers'      => [
+				'name'          => __( '[Newspack] Newsletter Subscribers', 'newspack-popups' ),
+				'configuration' => [
+					'is_subscribed' => true,
+				],
+			],
+			'everyone_else'    => [
+				'name'          => __( '[Newspack] Everyone Else', 'newspack-popups' ),
+				'configuration' => [
+					'is_not_donor'      => true,
+					'is_not_subscribed' => true,
+				],
+			],
+		];
+
+		foreach ( $default_options as $key => $segment ) {
+			// Ensure all config objects contain all option keys.
+			$segment['configuration'] = wp_parse_args( $segment['configuration'], self::DEFAULT_SEGMENT_CONFIG );
```

Comment:
`canonize_segment` didn't work for me. I saw two issues:

1. The object returned by that method is cast as a `stdClass Object` instead of an array, so that causes the segments to be saved to the options table in a way that can't be read by the Campaigns wizard UI.
2. Even if we cast the object to an array before saving, it doesn't seem to actually save the config settings—all segments are saved with default settings.

I think we should leave this as-is for now rather than refactor `canonize_segment` to be used here, as that method is pretty important for other features and I don't want to cause any regressions there.

Reply:
> Even if we cast the object to an array before saving, it doesn't seem to actually save the config settings—all segments are saved with default settings.

Casting to array _did_ work for me (also had to `require_once` the module at the top of the function, [like here](https://github.com/Automattic/newspack-popups/blob/debe5f6873c566ad801d25976a84fb2499552ab9/includes/class-newspack-popups-segmentation.php#L610)) –
 
```suggestion
			$segment['configuration'] = (array) Campaign_Data_Utils::canonize_segment( $segment['configuration'] );
```

I don't understand what's meant by "it doesn't seem to actually save the config settings" – the `Campaign_Data_Utils::canonize_segment` method is not supposed to save data, only process a configuration object.

Could you try again with the code here, unless it's exactly what you've already tried?

Reply:
Strange, I thought this is exactly what I tried, but I must have done something wrong because it's now working. Previously, when I tried to implement it would save all of the segments with default settings values. Anyway, fixed in 86337ae.